pkg/traceql/util.go

```diff
@@ -81,3 +83,61 @@ func (b *bucketSet) addAndTest(i int) bool {
 	b.buckets[b.sz]++
 	return false
 }
+
+const (
+	leftBranch  = 0
+	rightBranch = 1
+)
+
+type branchOptimizer struct {
+	start            time.Time
+	last             []time.Duration
+	totals           []time.Duration
+	Recording        bool
+	samplesRemaining int
+}
+
+func newBranchPredictor(numBranches int, numSamples int) branchOptimizer {
+	return branchOptimizer{
+		totals:           make([]time.Duration, numBranches),
+		last:             make([]time.Duration, numBranches),
+		samplesRemaining: numSamples,
+		Recording:        true,
+	}
+}
+
+// Start recording. Should be called immediately prior to a branch execution.
+func (b *branchOptimizer) Start() {
+	b.start = time.Now()
```

Comment:
I don't think this is a blocker, but I am concerned about the use of `time.Now()`. Since this is measuring system time and not "on cpu time" it could, in practice, record significantly larger values due to cpu contention.

What do we think about assigning a static integer weight to an operation? We could give integer comparisons very low weights, string comparison medium weights, regex comparisons high weights and extremely long regexes even higher weights. This would force the query to reorder itself to prefer cheaper operations w/o relying on system timing.

Obviously, it's not as dynamic as what you're proposing, but it does dodge some potential difficult to understand behavior in the future. 

To be clear I'm fine with this going in as is, but this one detail is picking at my brain.

Reply:
This is true, those are cases where the approach won't have accurate numbers.  Your brain is probably thinking of it because it's the approach in https://github.com/grafana/tempo/pull/4076   In that case the query-frontend doesn't have the benefit of execution time like we do here.  The worst case is that it chooses the slower path, but keep in mind that as of a few weeks ago before #4114 , it was always executing both paths. So performance will be somewhere less than ideal but still better than a few weeks ago.

The benefits of this approach I have noticed is that it is able to tease out small differences in operations that look similar on the surface, but give a few % boost by reordering.  

The weights approach is nice in that it's a static upfront, but I do think the tuning and brittleness of keeping up with the language would be a chore.

Reply:
yup, agreed.

> In that case the query-frontend doesn't have the benefit of execution time like we do here.

An interesting path on the frontend that I considered but didn't implement was to use job timing as way to dynamically weight requests in a batch. start with a small batch size and slowly increase the items until you hit some kind of ideal time for a full batch to be executed.

> I do think the tuning and brittleness of keeping up with the language would be a chore.

I'm concerned about this as well in the frontend weights PR you linked. It has a lot of potential but will require upkeep and attention to get the most value out of it. i was musing with @javiermolinar: if this PR, the frontend job weights PR you linked and a potential future JoinIterator reorder PR could all benefit from the concept of static weights maybe we could consolidate their calculation?

no easy answers here. let's plow forward with our best ideas. hopefully we'll look back in a few months and have stronger sense of how to pull off these optimizations. maybe we start leaning one path or the other?

thanks for the thoughts. code on!